File: C/TREE/Tree_11.c
/*
===============================================================================
Program Name    : Binary Search Tree - Count Parent Nodes
Program Number  : 011
File Name       : Tree_11.c

Description     : This program demonstrates how to count the
                  total number of Parent Nodes present in a
                  Binary Search Tree using recursion.

Date            : July 2026
Language        : C
Author          : Atharv Tushar Bhosale
===============================================================================

Concepts Used:
1. Self-Referential Structure
2. Dynamic Memory Allocation
3. Binary Search Tree (BST)
4. Recursion
5. Parent Node Counting

Key Learning Points:
- Create a Binary Search Tree.
- Insert nodes according to BST rules.
- Search elements in BST.
- Count total nodes.
- Count leaf nodes.
- Count parent nodes recursively.

===============================================================================
*/

// TREE

#include<stdio.h>
#include<stdlib.h>
#include<stdbool.h>

#pragma pack(1)
struct node
{
    int data;
    struct node *lchild;
    struct node *rchild;
};

typedef struct node NODE;
typedef struct node* PNODE;
typedef struct node** PPNODE;

// L    D   R
void Inorder(PNODE first)
{
    if(first != NULL)
    {
        Inorder(first -> lchild);
        printf("%d\n",first -> data);
        Inorder(first -> rchild);
    }
}

// D    L   R
void Preorder(PNODE first)
{
    if(first != NULL)
    {
        printf("%d\n",first -> data);
        Preorder(first -> lchild);
        Preorder(first -> rchild);
    }
}

// L    R   D
void Postorder(PNODE first)
{
    if(first != NULL)
    {
        Postorder(first -> lchild);
        Postorder(first -> rchild);
        printf("%d\n",first -> data);
    }
}

void Insert(PPNODE first, int iNo)
{
    PNODE newn = NULL;
    PNODE temp = NULL;

    newn = (PNODE)malloc(sizeof(NODE));

    newn -> data = iNo;
    newn -> lchild = NULL;
    newn -> rchild = NULL;

    if(*first == NULL)
    {
        *first = newn;
    }
    else
    {
        temp = *first;

        while(1)
        {
            if(iNo > temp -> data)
            {
                if(temp -> rchild == NULL)
                {
                    temp -> rchild = newn;
                    break;
                }
                temp = temp -> rchild;
            }
            else if(iNo < temp -> data)
            {
                if(temp -> lchild == NULL)
                {
                    temp -> lchild = newn;
                    break;
                }
                temp = temp -> lchild;
            }
            else if(iNo == temp -> data)
            {
                printf("Unable to Insert as Elements is Duplicate");
                free(newn);
                break;
            }
        }
    }
}

bool Search(PNODE first, int iNo)
{
    bool bFlag = false;

    while(first != NULL)
    {
        if(iNo == first -> data)
        {
            bFlag = true;
            break;
        }
        else if(iNo > first -> data)
        {
            first = first -> rchild;
        }
        else if(iNo < first -> data)
        {
            first = first -> lchild;
        }
    }
    return bFlag;
}

int Count(PNODE first)
{
    static int iCount = 0;
    

    if(first != NULL)
    {
        iCount++;
        Count(first -> lchild);
        Count(first -> rchild);
    }
    return iCount;
}

int CountLeaf(PNODE first)
{
    static int iCount = 0;

    if(first != NULL)
    {
        if(first -> lchild == NULL && first -> rchild == NULL)
        {
            iCount++;
        }
        CountLeaf(first -> lchild);
        CountLeaf(first -> rchild);
    }
    return iCount;
}

int CountParent(PNODE first)
{
    static int iCount = 0;

    if(first != NULL)
    {
        if(first -> lchild != NULL || first -> rchild != NULL)
        {
            iCount++;
        }
        CountParent(first -> lchild);
        CountParent(first -> rchild);
    }
    return iCount;
}

int main()
{
    PNODE head = NULL;
    int iRet = 0;

    Insert(&head,11);
    Insert(&head,5);
    Insert(&head,17);
    Insert(&head,21);
    Insert(&head,4);
    Insert(&head,7);
    Insert(&head,15);

    printf("Inorder Display: \n");
    Inorder(head);

    iRet = Count(head);
    printf("Number of Nodes are: %d\n",iRet);

    if(Search(head,25) == true)
    {
        printf("25 is present in BST\n");
    }
    else
    {
        printf("25 is not present in BST\n");
    }
    
    iRet = CountLeaf(head);
    printf("Number of Leaf Nodes are: %d\n",iRet);

    iRet = CountParent(head);
    printf("Number of Parent Nodes are: %d\n",iRet);

    return 0;
}

/*
===============================================================================
Expected Output:
===============================================================================

Inorder Display:

4
5
7
11
15
17
21

Number of Nodes are: 7

25 is not present in BST

Number of Leaf Nodes are: 4

Number of Parent Nodes are: 3

===============================================================================
Memory Representation:
===============================================================================

                11
              /    \
             5      17
            / \    /  \
           4   7  15  21

Parent Nodes : 11, 5, 17
Leaf Nodes   : 4, 7, 15, 21

===============================================================================
Algorithm:
===============================================================================

1. Start from root node.
2. Check whether current node has at least one child.
3. If left child or right child exists, increment count.
4. Recursively traverse left subtree.
5. Recursively traverse right subtree.
6. Return total parent node count.

===============================================================================
Condition Used:
===============================================================================

if(first->lchild != NULL || first->rchild != NULL)

A Parent Node is a node having at least one child.

===============================================================================
Time Complexity:
===============================================================================

Insert()       : O(log n) Average
Search()       : O(log n) Average

Count()        : O(n)

CountLeaf()    : O(n)

CountParent()  : O(n)

Space Complexity : O(h)

where,

n = Number of Nodes
h = Height of Tree

===============================================================================
Conclusion:
===============================================================================

Program 009 demonstrates how to count Parent Nodes in a
Binary Search Tree using recursion. A Parent Node is a node
that contains at least one child node. Each node is visited
exactly once during traversal.

===============================================================================
*/

File: Java/Tree/Tree_11.java
/*
===============================================================================
Program Name    : Binary Search Tree - Menu Driven Application
Program Number  : 011
File Name       : Tree_11.java

Description     : This program demonstrates various Binary
                  Search Tree operations using a menu-driven
                  application.

                  Operations:
                  1. Insert Node
                  2. Search Node
                  3. Inorder Traversal
                  4. Preorder Traversal
                  5. Postorder Traversal
                  6. Count Total Nodes
                  7. Count Leaf Nodes
                  8. Count Parent Nodes
                  0. Exit

Concepts Used:
1. Class and Object
2. Self-Referential Class
3. Binary Search Tree
4. Dynamic Memory Allocation
5. Recursion
6. Tree Traversals
7. Searching
8. Counting Nodes
9. Menu Driven Programming

Date            : August 2026
Language        : Java
Author          : Atharv Tushar Bhosale
===============================================================================
*/

import java.util.*;

class node
{
    public int data;
    public node lchild;
    public node rchild;
}

class Tree
{
    private node first;

    public Tree()
    {
        first = null;
    }

    ///////////////////////////////////////////////////////////////////////////
    //
    // Function Name : Insert
    // Description   : Insert element into Binary Search Tree
    //
    ///////////////////////////////////////////////////////////////////////////

    public void Insert(int iNo)
    {
        node newn = null;
        node temp = null;

        newn = new node();

        newn.data = iNo;
        newn.lchild = null;
        newn.rchild = null;

        if(first == null)
        {
            first = newn;
        }
        else
        {
            temp = first;

            while(true)
            {
                if(iNo > temp.data)
                {
                    if(temp.rchild == null)
                    {
                        temp.rchild = newn;
                        break;
                    }

                    temp = temp.rchild;
                }
                else if(iNo < temp.data)
                {
                    if(temp.lchild == null)
                    {
                        temp.lchild = newn;
                        break;
                    }

                    temp = temp.lchild;
                }
                else
                {
                    System.out.println("Duplicate element not allowed");
                    break;
                }
            }
        }
    }

    ///////////////////////////////////////////////////////////////////////////
    //
    // Function Name : Search
    // Description   : Search element in Binary Search Tree
    //
    ///////////////////////////////////////////////////////////////////////////

    public boolean Search(int iNo)
    {
        node temp = null;

        temp = first;

        while(temp != null)
        {
            if(iNo == temp.data)
            {
                return true;
            }
            else if(iNo > temp.data)
            {
                temp = temp.rchild;
            }
            else
            {
                temp = temp.lchild;
            }
        }

        return false;
    }

    ///////////////////////////////////////////////////////////////////////////
    //
    // Function Name : Inorder
    // Description   : Left -> Data -> Right
    //
    ///////////////////////////////////////////////////////////////////////////

    public void Inorder(node temp)
    {
        if(temp != null)
        {
            Inorder(temp.lchild);
            System.out.println(temp.data);
            Inorder(temp.rchild);
        }
    }

    public void Inorder()
    {
        Inorder(first);
    }

    ///////////////////////////////////////////////////////////////////////////
    //
    // Function Name : Preorder
    // Description   : Data -> Left -> Right
    //
    ///////////////////////////////////////////////////////////////////////////

    public void Preorder(node temp)
    {
        if(temp != null)
        {
            System.out.println(temp.data);
            Preorder(temp.lchild);
            Preorder(temp.rchild);
        }
    }

    public void Preorder()
    {
        Preorder(first);
    }

    ///////////////////////////////////////////////////////////////////////////
    //
    // Function Name : Postorder
    // Description   : Left -> Right -> Data
    //
    ///////////////////////////////////////////////////////////////////////////

    public void Postorder(node temp)
    {
        if(temp != null)
        {
            Postorder(temp.lchild);
            Postorder(temp.rchild);
            System.out.println(temp.data);
        }
    }

    public void Postorder()
    {
        Postorder(first);
    }

    ///////////////////////////////////////////////////////////////////////////
    //
    // Function Name : Count
    // Description   : Count total number of nodes
    //
    ///////////////////////////////////////////////////////////////////////////

    public int Count(node temp)
    {
        if(temp == null)
        {
            return 0;
        }

        return Count(temp.lchild) +
               Count(temp.rchild) + 1;
    }

    public int Count()
    {
        return Count(first);
    }

    ///////////////////////////////////////////////////////////////////////////
    //
    // Function Name : CountLeaf
    // Description   : Count leaf nodes
    //
    ///////////////////////////////////////////////////////////////////////////

    public int CountLeaf(node temp)
    {
        if(temp == null)
        {
            return 0;
        }

        if(temp.lchild == null && temp.rchild == null)
        {
            return 1;
        }

        return CountLeaf(temp.lchild) +
               CountLeaf(temp.rchild);
    }

    public int CountLeaf()
    {
        return CountLeaf(first);
    }

    ///////////////////////////////////////////////////////////////////////////
    //
    // Function Name : CountParent
    // Description   : Count parent nodes
    //
    ///////////////////////////////////////////////////////////////////////////

    public int CountParent(node temp)
    {
        if(temp == null)
        {
            return 0;
        }

        if(temp.lchild != null || temp.rchild != null)
        {
            return 1 +
                   CountParent(temp.lchild) +
                   CountParent(temp.rchild);
        }

        return 0;
    }

    public int CountParent()
    {
        return CountParent(first);
    }
}

class Tree_11
{
    public static void main(String A[])
    {
        Scanner sobj = new Scanner(System.in);

        Tree tobj = new Tree();

        int iChoice = 0;
        int iNo = 0;

        while(true)
        {
            System.out.println("\n==============================================");
            System.out.println("       BINARY SEARCH TREE MENU");
            System.out.println("==============================================");

            System.out.println("1. Insert Node");
            System.out.println("2. Search Node");
            System.out.println("3. Inorder Traversal");
            System.out.println("4. Preorder Traversal");
            System.out.println("5. Postorder Traversal");
            System.out.println("6. Count Total Nodes");
            System.out.println("7. Count Leaf Nodes");
            System.out.println("8. Count Parent Nodes");
            System.out.println("0. Exit");

            System.out.println("==============================================");

            System.out.print("Enter your choice : ");
            iChoice = sobj.nextInt();

            switch(iChoice)
            {
                case 1:

                    System.out.print("Enter element : ");
                    iNo = sobj.nextInt();

                    tobj.Insert(iNo);

                    break;

                case 2:

                    System.out.print("Enter element to search : ");
                    iNo = sobj.nextInt();

                    if(tobj.Search(iNo) == true)
                    {
                        System.out.println("Element is Present");
                    }
                    else
                    {
                        System.out.println("Element is Not Present");
                    }

                    break;

                case 3:

                    System.out.println("\nInorder Traversal :");

                    tobj.Inorder();

                    break;

                case 4:

                    System.out.println("\nPreorder Traversal :");

                    tobj.Preorder();

                    break;

                case 5:

                    System.out.println("\nPostorder Traversal :");

                    tobj.Postorder();

                    break;

                case 6:

                    System.out.println("Total Nodes : " + tobj.Count());

                    break;

                case 7:

                    System.out.println("Leaf Nodes : " + tobj.CountLeaf());

                    break;

                case 8:

                    System.out.println("Parent Nodes : " + tobj.CountParent());

                    break;

                case 0:

                    System.out.println("\nThank you for using Binary Search Tree Application");

                    System.exit(0);

                    break;

                default:

                    System.out.println("Invalid Choice");

                    break;
            }
        }
    }
}

/*
===============================================================================
Sample Input / Output

==============================================
       BINARY SEARCH TREE MENU
==============================================

1. Insert Node
2. Search Node
3. Inorder Traversal
4. Preorder Traversal
5. Postorder Traversal
6. Count Total Nodes
7. Count Leaf Nodes
8. Count Parent Nodes
0. Exit

Enter your choice : 1
Enter element : 11

Enter your choice : 1
Enter element : 5

Enter your choice : 1
Enter element : 17

Enter your choice : 1
Enter element : 4

Enter your choice : 1
Enter element : 7

Enter your choice : 1
Enter element : 15

Enter your choice : 1
Enter element : 21

Enter your choice : 3

Inorder Traversal :
4
5
7
11
15
17
21

Enter your choice : 4

Preorder Traversal :
11
5
4
7
17
15
21

Enter your choice : 5

Postorder Traversal :
4
7
5
15
21
17
11

Enter your choice : 6
Total Nodes : 7

Enter your choice : 7
Leaf Nodes : 4

Enter your choice : 8
Parent Nodes : 3

Enter your choice : 2
Enter element to search : 15
Element is Present

Enter your choice : 0

Thank you for using Binary Search Tree Application

===============================================================================
Tree Structure

              11
            /    \
           5      17
          / \    /  \
         4   7  15  21

===============================================================================
Time Complexity

Insert()          : O(log n) Average
Search()          : O(log n) Average
Inorder()         : O(n)
Preorder()        : O(n)
Postorder()       : O(n)
Count()           : O(n)
CountLeaf()       : O(n)
CountParent()     : O(n)

Worst Case for Insert/Search : O(n)

Space Complexity : O(h)

where h = height of the tree

===============================================================================
End of Program
===============================================================================
*/
